### ISSUE TITLE
[Route53] Spuriously REFUSES to create hosted zone multiple times

### ISSUE THREAD

**Issue by axmetishe** [saltstack/salt#40655]
### Description of Issue/Question
Currently boto_route53 create_hosted_zone state use property name to pass it as caller reference.
This leads to error "A hosted zone has already been created with the specified caller reference." in case when you create hosted zone, then delete it and tried to create again.

From Boto documentation:
> CallerReference (string) --
[REQUIRED]
A unique string that identifies the request and that allows failed CreateHostedZone requests to be retried without the risk of executing the operation twice. You must use a unique CallerReference string every time you create a hosted zone. CallerReference can be any unique string, for example, a date/time stamp.

Basically to avoid duplication in case while boto_module run several retries we should provide exactly the same caller_ref only for execution time, so we can pass via state generated uuid and it will not be changed during execution time and will prevent duplication.

My suggestion is unmap caller_ref from name parameter in state and pass just uuid4 as used in boto connection module.

What you guys think about that?

### Setup
```yaml
manage '{{ zone }}' hosted zone:
  boto_route53.hosted_zone_present:
    - name: {{ zone }}
    - domain_name: {{ zone }}
    - comment: {{ parameters.comment }}
    - private_zone: {{ parameters.private }}
    - profile: {{ configurations.profile }}
```

### Steps to Reproduce Issue
Create hosted zone then remove and create again.

```
[INFO    ] Running state [dev.example.com.] at time 08:38:07.433090
[INFO    ] Executing state boto_route53.hosted_zone_present for dev.example.com.
[ERROR   ] Failed to create hosted zone dev.example.com.: A hosted zone has already been created with the specified caller reference.
[ERROR   ] Creating Hosted Zone dev.example.com. failed
[INFO    ] Completed state [dev.example.com.] at time 08:38:09.145868 duration_in_ms=1712.778
```

### Versions Report
```
Salt Version:
           Salt: 2016.11.2
 
Dependency Versions:
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.6.0
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.21.0
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: 0.21.4
         Python: 2.7.5 (default, Nov 20 2015, 02:00:19)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.7.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: centos 7.2.1511 Core
        machine: x86_64
        release: 3.10.0-514.2.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.2.1511 Core
 ```

### Quick and dirty solution patch for testing
```
Index: salt/states/boto_route53.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- salt/states/boto_route53.py	(revision cb9dcb1e1b860b919cbdd5ea5c9273719bb2fdde)
+++ salt/states/boto_route53.py	(revision )
@@ -74,6 +74,7 @@
 # Import Python Libs
 from __future__ import absolute_import
 import json
+import uuid
 
 # Import Salt Libs
 from salt.utils import SaltInvocationError, exactly_one
@@ -359,7 +360,7 @@
     return ret
 
 
-def hosted_zone_present(name, domain_name=None, private_zone=False, comment='',
+def hosted_zone_present(name=None, domain_name=None, private_zone=False, comment='',
                         vpc_id=None, vpc_name=None, vpc_region=None,
                         region=None, key=None, keyid=None, profile=None):
     '''
@@ -481,7 +482,7 @@
             ret['result'] = None
             return ret
         res = __salt__['boto_route53.create_hosted_zone'](domain_name=domain_name,
-                caller_ref=name, comment=comment, private_zone=private_zone,
+                caller_ref=str(uuid.uuid4()), comment=comment, private_zone=private_zone,
                 vpc_id=vpc_id, vpc_region=vpc_region, region=region, key=key,
                 keyid=keyid, profile=profile)
         if res:
```

**Comment by Ch3LL** [comment]
@tkwilliams looks like you added this method. Do you have any suggestions for @axmetishe 's fix?

**Comment by tkwilliams** [comment]
Sorry, on vacation this week :)

It's been awhile and I don't have the source open in front of me at this exact moment, but IIRC I tried something like the above and ran into an issue because generating a UUID for each run caused a different "zone already exists" type error.  Because CallerReference was different at each run, AWS would try to recreate (e.g. not be idempotent) the zone at each run, naturally failing each one after the first.  I didn't actually TEST the "create-delete-recreate" flow you describe above, so I can't say I hit this exact issue, but I can easily believe AWS would do something like this based on similar behaviour I've seen elsewhere.

I'm assuming (again, based on behaviour elsewhere in AWS) that there is some reasonable timeout for such CallerReverence re-use, meaning that short-cycled "create-delete-create" will trigger an error, while longer (likely over 24 hours is quite safe) will not.  This is backed up by my past ability to delete and re-create zones, with the existing code -- but admittedly over the course of days, not minutes or hours.

If the salt state itself verifies the existence of the appropriately named zone file (and presumably its attributes) then idempotency will simply have to be handled entirely within salt, rather than giving AWS any real say in the matter - this makes the salt code heavier, and of course will incite more traffic back-and-forth with the API, but given the case you're hitting I really can't see any better approach than the one you suggest -- which is essentially to entirely short-circuit / bypass the CallerRef functionality by passing in a new value at every run.

I'll be back in the office Thur (Sep 28) and I'll try to give the function a quick peek then :)

**Comment by tkwilliams** [comment]
Looking at what I wrote for the boto3_route53 module, I seem to have reached the same conclusion as you :)

The boto3 based version in there uses `uuid.uuid4()` to generate CallerReference if one is not explicitly provided in the state definition.  That said, the DOCSTRING I seem to have copy-pasted from the original boto2 version and never changed, so it still talks about using the value of `name` even though the code implements as stated above.

You may want to look at boto3_route53 in any event, as it's much improved over the original module, with all the bells and whistles that only boto3 can provide :)

Just opened https://github.com/saltstack/salt/pull/43985 to address the concerns above, including the inaccurate docstring in boto3_route53.

**Comment by tkwilliams** [comment]
Actually, this ticket is mis-titled - it should be "[Route53] Spuriously REFUSES to create hosted zone multiple times"

**Comment by Ch3LL** [comment]
@axmetishe can you give the fix in #43985 a try?

**Comment by tkwilliams** [comment]
@Ch3LL  Per the discussion with @ryan-lane in #43985, I'm not sure that "fixing" it is actually the correct approach in any event.  Let me finish up and PR these boto3 ElasticMapReduce modules I'm working on (another day or so max) and I'll revisit the above fix PR with (hopefully) some code which will resolve BOTH issues.  As it stands, the fix I submitted simply trades one problem for another...

@axmetishe  Just for clarity, would you mind renaming this ticket to something like "[Route53] Spuriously REFUSES to create hosted zone multiple times", which is a more accurate description and will help people when searching in the future mebbe :)

**Comment by axmetishe** [comment]
@Ch3LL these changes are exactly the same as in my patch, so yes it will work, moreover we are using this fix for more than 6 months in my company and all is working as expected.

@tkwilliams I don't really worry about hosted zone duplication, because AWS API allow to create any hosted zone even not owned by you, for example I could create in my account github.com, but it will work only for me and for whom will use AWS DNS servers provided from my this hosted zone, so basically AWS API allow domain duplication. On the other hand we have boto route53 module with method `create_hosted_zone` which have `caller_ref` parameter: 

> caller_ref (str) – A unique string that identifies the request and that allows failed CreateHostedZone requests to be retried without the risk of executing the operation twice. If you don’t provide a value for this, boto will generate a Type 4 UUID

Same parameter required for boto3 module:
> CallerReference (string) --
[REQUIRED]
A unique string that identifies the request and that allows failed CreateHostedZone requests to be retried without the risk of executing the operation twice. You must use a unique CallerReference string every time you submit a CreateHostedZone request. CallerReference can be any unique string, for example, a date/time stamp.

So basically even if we will use UUID it will not try to create duplicated hosted zone, because we still checking hosted zone for existence via name, tags. Apparently only one situation as I can see could fail execution - multi-master configuration, for example when state will be running on two of them in some cases we could get race condition, this should be checked, because both of them could actually normally handle creation due to existence check.

P.S. sorry for late answer.
P.P.S.  @tkwilliams I'll rename the issue.

**Comment by tkwilliams** [comment]
Circling back through these older tickets....
To the best of my knowledge, this was resolved by  https://github.com/saltstack/salt/pull/43985

@axmetishe, can you verify (or it's opposite if appropriate) that current salt (which contains the above PR) now does what you need, so we can close out the issue?

Thanks :)
t.

**Comment by tkwilliams** [comment]
@Ch3LL  Is it safe to close this one out?  I'm pretty confident it was fixed in the above cited PR.

**Comment by Ch3LL** [comment]
Yep i'll go ahead and close and if anyone needs it re-opened just let me know
